pennylane/ops/op_math/linear_combination.py

```diff
@@ -0,0 +1,503 @@
+    >>> obs = [qml.X(0) @ qml.Z(1), qml.Z(0) @ qml.Hadamard(2)]
+    >>> H = qml.ops.LinearCombination(coeffs, obs)
+    >>> print(H)
+    0.2 * (X(0) @ Z(1)) + -0.543 * (Z(0) @ Hadamard(wires=[2]))
+
+
+    A LinearCombination can store information on which commuting observables should be measured together in
+    a circuit:
+
+    >>> obs = [qml.X(0), qml.X(1), qml.Z(0)]
+    >>> coeffs = np.array([1., 2., 3.])
+    >>> H = qml.ops.LinearCombination(coeffs, obs, grouping_type='qwc')
+    >>> H.grouping_indices
+    ((0, 1), (2,))
+
+    This attribute can be used to compute groups of coefficients and observables:
+
+    >>> grouped_coeffs = [coeffs[list(indices)] for indices in H.grouping_indices]
+    >>> grouped_obs = [[H.ops[i] for i in indices] for indices in H.grouping_indices]
+    >>> grouped_coeffs
+    [array([1., 2.]), array([3.])]
+    >>> grouped_obs
+    [[X(0), X(1)], [Z(0)]]
+
+    Devices that evaluate a LinearCombination expectation by splitting it into its local observables can
+    use this information to reduce the number of circuits evaluated.
+
+    Note that one can compute the ``grouping_indices`` for an already initialized LinearCombination by
+    using the :func:`compute_grouping <pennylane.LinearCombination.compute_grouping>` method.
+    """
+
+    num_wires = qml.operation.AnyWires
+    grad_method = "A"  # supports analytic gradients
+    batch_size = None
+    ndim_params = None  # could be (0,) * len(coeffs), but it is not needed. Define at class-level
+
+    def _flatten(self):
+        # note that we are unable to restore grouping type or method without creating new properties
+        return (self._coeffs, self._ops, self.data), (self.grouping_indices,)
+
+    @classmethod
+    def _unflatten(cls, data, metadata):
+        new_op = cls(data[0], data[1])
+        new_op._grouping_indices = metadata[0]  # pylint: disable=protected-access
+        new_op.data = data[2]
+        return new_op
+
+    def __init__(
+        self,
+        coeffs,
+        observables: List[Operator],
+        simplify=False,
+        grouping_type=None,
+        method="rlf",
+        _pauli_rep=None,
+        id=None,
+    ):
+        if qml.math.shape(coeffs)[0] != len(observables):
+            raise ValueError(
+                "Could not create valid LinearCombination; "
+                "number of coefficients and operators does not match."
+            )
+
+        self._coeffs = coeffs
+
+        self._ops = [convert_to_opmath(op) for op in observables]
+
+        self._hyperparameters = {"ops": self._ops}
+
+        self._grouping_indices = None
+
+        with qml.QueuingManager().stop_recording():
+            operands = [qml.s_prod(c, op) for c, op in zip(coeffs, observables)]
+
+        super().__init__(
+            *operands, grouping_type=grouping_type, method=method, id=id, _pauli_rep=_pauli_rep
+        )
+
+        if simplify:
+            # TODO clean up this logic, seems unnecesssarily complicated
```

Comment:
It would be easier if we make certain functions static, like `build_pauli_rep` and `simplify_coeffs_ops`, this way we don't have to initialize the entire object before simplifying it. Instead we would be able to do something like
```
if simplify:
    coeffs, ops, pr = _simplify_coeffs_ops(coeffs, ops, pr)
```
before initializing everything else, this way we won't have duplicate code like this.

Reply:
I second this.

Reply:
Very nice idea!
Because it is a bigger change I opened a separate PR here https://github.com/PennyLaneAI/pennylane/pull/5419
Currently not yet fully working because I get an odd opposite sign when differentiating and using `simplify=True`